pybmix/core/pybmixcpp/python_hierarchy.cc

```diff
@@ -0,0 +1,233 @@
+    int size = statecast.general_state().size();
+    std::vector<double> aux_v{};
+    for (int i = 0; i < size; ++i) {
+        aux_v.push_back((statecast.general_state().data())[i]);
+    }
+    state.generic_state = aux_v;
+    set_card(statecast.cardinality());
+}
+
+//! C++
+std::shared_ptr <bayesmix::AlgorithmState::ClusterState>
+PythonHierarchy::get_state_proto() const {
+    bayesmix::Vector state_;
+    state_.set_size(state.generic_state.size());
+    *state_.mutable_data() = {
+            state.generic_state.data(),
+            state.generic_state.data() + state.generic_state.size()};
+    auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+    out->mutable_general_state()->CopyFrom(state_);
+    return out;
+}
+
+//! C++
+void PythonHierarchy::set_hypers_from_proto(
+        const google::protobuf::Message &hypers_) {
+    auto &hyperscast = downcast_hypers(hypers_).python_state();
+    int size = hyperscast.data().size();
+    std::vector<double> aux_v{};
+    for (int i = 0; i < size; ++i) {
+        aux_v.push_back((hyperscast.data())[i]);
+    }
+    hypers->generic_hypers = aux_v;
+}
+
+//! C++
+std::shared_ptr <bayesmix::AlgorithmState::HierarchyHypers>
+PythonHierarchy::get_hypers_proto() const {
+    bayesmix::Vector hypers_;
+    hypers_.set_size(hypers->generic_hypers.size());
+    *hypers_.mutable_data() = {
+            hypers->generic_hypers.data(),
+            hypers->generic_hypers.data() + hypers->generic_hypers.size()};
+    auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+    out->mutable_python_state()->CopyFrom(hypers_);
+    return out;
+}
+
+void synchronize_cpp_to_py_state(const std::mt19937 &cpp_gen,
+                                 py::object &py_gen) {
+    std::stringstream state{};
+    state << cpp_gen;
+
+    std::string aux_string{};
+    py::list state_list{};
+
+    for (unsigned int n = 0; n < 624; ++n) {
+        state >> aux_string;
+        state_list.append(std::stoul(aux_string));
+    }
+    state >> aux_string;
+    unsigned int pos = std::stoul(aux_string);
+
+    py::object array = numpy.attr("array")(state_list, "uint32");
+    py::dict state_dict("key"_a = array, "pos"_a = pos);
+    py::dict d("bit_generator"_a = "MT19937", "state"_a = state_dict);
+    py_gen.attr("__setstate__")(d);
+}
+
+void synchronize_py_to_cpp_state(std::mt19937 &cpp_gen,
+                                 const py::object &py_gen) {
+    py::object py_state = py_gen.attr("__getstate__")();
+    py::object state_ = py_state["state"]["key"].attr("tolist")();
+    auto pos_ = py_state["state"]["pos"].cast<unsigned int>();
+
+    std::stringstream ss_state_;
+    for (auto elem: state_) {
+        ss_state_ << elem.cast<unsigned int>() << " ";
+    }
+    ss_state_ << pos_;
+    ss_state_ >> cpp_gen;
+}
+
+py::list vector_to_list(const std::vector<double> &x) {
+    py::list L;
+    for (auto &elem: x) {
+        L.append(elem);
+    }
+    return L;
```

Comment:
I would believe that there is a faster way than appending, perhaps already reserving the correct size?

Reply:
This might help: https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html

Reply:
Thank you. We had changed the same thing in another function but hadn't noticed here. It affected a lot apparently. Now it runs much faster both with Neal2,3. 1100 iterations in 97 seconds for Neal3.

Reply:
Thank you. We modified the code to use the pybind11 built-in type conversions.
We will run the profiler in the upcoming days and will share the results.

About the make_opaque, we're still in the reading phase :)

Reply:
Hi,
we tried experimenting with MAKE_OPAQUE to pass the state vector by reference.

We turned std::vector into an opaque type (called DoubleVector) and managed to implement the push_back method in the corresponding class declaration. When overloading the access operator [] however we encountered the first problem: it worked for elements outside the vector as well, returning random numbers.

Moreover, since MAKE_OPAQUE deactivates the automatic conversion between python lists and std::vector we changed the state data structure into a generic pybind11 object, which would then be passed a DoubleVector object. We encountered some problems however in passing it by reference to python.

Finally it appears to us that most of the pybind11 documentation deals with running c++ from python, and not viceversa (as in our case). Can you point us towards some working examples of MAKE_OPAQUE related to what we're trying to do?

Reply:
I am not an expert on this. One option would be to open an issue in the `pybind11` github and see if a developer answers!

Reply:
Hi, after receiving the answer on the pybind11 github we were finally able to make it work. However, it doesn't make the code any faster (actually it seems to slightly slow it down). This is probably because the vectors we are passing are of extremely small size. 
You can find the implementation in the [py_hier_opaque](https://github.com/taguhiM/pybmix/tree/py_hier_opaque) branch, is it better to keep this approach (which maybe will make code faster in the event someone needs to pass bigger vectors) or to use the automatic copy conversions?

Reply:
Ok, that's interesting!

I don't have a clear answer to that, maybe the best thing is to benchmark the approaches on a multivariate case. We're not interested in anything fancy.
I would do a mock-up where you don't code all the stuff properly. E.g. the lpdf evaluator might return simply the sum of data (even if this does not make any sense) and the sampler might just sampler randomly from a normal distribution without any meaning. Just to benchmark the conversion speed!